### PR TITLE
feat: Auto infra sync after 7 days

### PR DESCRIPTION
**Pull request by hnnasit**
#### Which issue(s) does this change fix?

N/A

#### Why is this change necessary?
To minimize the drift in CloudFormation due to only running sync flows to update the code resources. 

#### How does it address the issue?
The logic will store the last sync time after every infra sync and run an infra sync if it detects an infra sync hasn't been done in a week.

#### What side effects does this change have?
N/A

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/commands/sync/sync_context.py [hunk 1]
```diff
@@ -39,11 +40,29 @@ class ResourceSyncState:
 @dataclass
 class SyncState:
     dependency_layer: bool
+    latest_infra_sync_time: Optional[datetime]
     resource_sync_states: Dict[str, ResourceSyncState]
 
     def update_resource_sync_state(self, resource_id: str, hash_value: str):
```
Comment (qingchm) [inline samcli/commands/sync/sync_context.py:46]:
```suggestion
    def update_resource_sync_state(self, resource_id: str, hash_value: str) -> None:
```

File: samcli/commands/sync/sync_context.py [hunk 2]
```diff
@@ -39,11 +40,29 @@ class ResourceSyncState:
...
+        Parameters
+        -------
+        resource_id: str
+            The resource identifier of the resource
+        hash_value: str
+            The logical ID identifier of the resource
+        """
         self.resource_sync_states[resource_id] = ResourceSyncState(hash_value, datetime.utcnow())
 
+    def update_infra_sync_time(self):
```
Comment (qingchm) [inline samcli/commands/sync/sync_context.py:60]:
```suggestion
    def update_infra_sync_time(self) -> None:
```

File: samcli/lib/sync/infra_sync_executor.py [hunk 3]
```diff
@@ -179,8 +196,18 @@ def execute_infra_sync(self, first_sync: bool = False) -> InfraSyncResult:
             #     )
             pass
 
+        # Will be added with the sync infra skip is ready for release
+        # if days_since_last_infra_sync > AUTO_INFRA_SYNC_DAYS:
```
Comment (mndeveci) [inline samcli/lib/sync/infra_sync_executor.py:200]: I wonder if we can simplify this double if blocks over same condition. Not blocking now but something for GA PR.

**Review by mndeveci — APPROVED** [review]

File: samcli/lib/sync/infra_sync_executor.py (hunk 3 above)
Comment (hnnasit) [reply samcli/lib/sync/infra_sync_executor.py:200]: Yeah that makes sense. This if loop can be switched to an elif condition with the above if condition.

**Review by qingchm — APPROVED** [review]
Overall looks good, made small comments on debug message improvements and dataclass field order which the next commit will address

**Pull request merged**